### PR TITLE
Completed callbacks

### PR DESCRIPTION
**Pull request by abhikanojia**

### CODE REVIEW

File: depot/test/mailers/user_mailer_test.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (harsh-ag) [inline depot/test/mailers/user_mailer_test.rb]: remove unused files.

File: depot/test/mailers/previews/user_mailer_preview.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Preview all emails at http://localhost:3000/rails/mailers/user_mailer
```
Comment (harsh-ag) [inline depot/test/mailers/previews/user_mailer_preview.rb]: remove unused files.

File: depot/app/controllers/users_controller.rb [hunk 3]
```diff
@@ -79,4 +83,10 @@ def ensure_user_not_destroying_itself
         redirect_to users_url, notice: "Cannot delete yourself."
       end
     end
+
+    def verify_user_is_not_admin
```
Comment (harsh-ag) [inline depot/app/controllers/users_controller.rb]: what is the use of this method.

File: depot/app/controllers/users_controller.rb [hunk 4]
```diff
@@ -56,10 +55,15 @@ def update
...
-      format.html { redirect_to users_url, notice: 'User was successfully destroyed.' }
-      format.json { head :no_content }
+    if @user.destroy
+      respond_to do |format|
+        format.html { redirect_to users_url, notice: 'User was successfully destroyed.' }
+        format.json { head :no_content }
+      end
+    else
+      respond_to do |format|
+        format.html { redirect_to users_url, notice: 'Cannot delete admin.' }
```
Comment (harsh-ag) [inline depot/app/controllers/users_controller.rb]: there can be any other reason for user.destroy to return false.

File: depot/app/mailers/user_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class UserMailer < ApplicationMailer
+  default from: 'Abhishek Kanojia <abhikanojiatest@gmail.com>'
```
Comment (harsh-ag) [inline depot/app/mailers/user_mailer.rb]: - use constants
- will you write defualt from in each mailer?

File: depot/app/models/user.rb [hunk 6]
```diff
@@ -5,6 +5,9 @@ class User < ApplicationRecord
   has_secure_password
 
   after_destroy :ensure_an_admin_remains
+  before_destroy :ensure_not_destroying_admin
+  before_update :ensure_not_updating_admin
+  after_create :send_welcome_email_to_user
```
Comment (harsh-ag) [inline depot/app/models/user.rb]: use better callback over here.

File: depot/app/models/user.rb [hunk 7]
```diff
@@ -15,4 +18,17 @@ def ensure_an_admin_remains
         raise Error.new "Can't delete last user"
       end
     end
+
+    def send_welcome_email_to_user
+      UserMailer.welcome_user(self).deliver_now
+    end
+
+    def ensure_not_destroying_admin
+    throw :abort if self.email.eql? 'admin@depot.com'
```
Comment (harsh-ag) [inline depot/app/models/user.rb]: indentation

File: depot/app/models/user.rb [hunk 8]
```diff
@@ -15,4 +18,17 @@ def ensure_an_admin_remains
...
+    end
+
+    def ensure_not_updating_admin
+      errors.add(:base, 'Cannot update depot admin.') if self.email.eql? 'admin@depot.com'
```
Comment (harsh-ag) [inline depot/app/models/user.rb]: is this working correct

File: depot/app/models/user.rb (hunk 7 above)
Comment (harsh-ag) [inline depot/app/models/user.rb]: self.email.eql? 'admin@depot.com' is being used many times. create a method.

File: depot/app/models/product.rb [hunk 9]
```diff
@@ -20,6 +23,14 @@ class Product < ApplicationRecord
 
   private
 
+  def set_default_discount_price
+    self.discount_price = price
+  end
+
+  def set_default_title
+    self.title = 'abc'
```
Comment (harsh-ag) [inline depot/app/models/product.rb]: constant.

File: depot/app/controllers/users_controller.rb [hunk 10]
```diff
@@ -1,8 +1,7 @@
 class UsersController < ApplicationController
   include UserLoggedIn
   before_action :set_user, only: [:show, :edit, :update, :destroy]
-  before_action :ensure_user_not_destroying_itself, only: [:destroy]
-
+  before_action :ensure_user_not_destroying_itself, only: :destroy
```
Comment (harsh-ag) [inline depot/app/controllers/users_controller.rb]: this logic should be in model.

**Review by harsh-ag — CHANGES_REQUESTED** [review]

File: depot/test/mailers/user_mailer_test.rb (hunk 1 above)
Comment (abhikanojia) [reply depot/test/mailers/user_mailer_test.rb]: ok

File: depot/app/models/user.rb (hunk 8 above)
Comment (abhikanojia) [reply depot/app/models/user.rb]: yes

File: depot/app/controllers/users_controller.rb (hunk 3 above)
Comment (abhikanojia) [reply depot/app/controllers/users_controller.rb]: removed

File: depot/app/controllers/users_controller.rb (hunk 10 above)
Comment (abhikanojia) [reply depot/app/controllers/users_controller.rb]: this function is checking whether user is not deleting a user by which he/she is logged in to application.

File: depot/app/models/user.rb [hunk 11]
```diff
@@ -15,4 +18,23 @@ def ensure_an_admin_remains
         raise Error.new "Can't delete last user"
       end
     end
+
+    def send_welcome_email_to_user
+      UserMailer.welcome_user(self).deliver_now
+    end
+
+    def user_is_admin?
```
Comment (harsh-ag) [inline depot/app/models/user.rb:26]: rename. you are writing code in user model so no need to add user in method.

File: depot/app/views/errors/_error.html.erb [hunk 12]
```diff
@@ -1,10 +1,9 @@
...
   <div id="error_explanation">
-    <h2><%=raw t('errors.error.header', count: @order.errors.count,
-        model: t('activerecord.models.order')) %>.</h2>
-    <p><%= t('errors.error.body') %></p>
+    <h2><%=raw t('errors.error.header', count: object.errors.count,
+        model: t('activerecord.modelsobject')) %>.</h2>
     <ul>
-    <% @order.errors.full_messages.each do |message| %>
+    <% object.errors.full_messages.each do |message| %>
       <li><%=raw message %></li>
```
Comment (harsh-ag) [inline depot/app/views/errors/_error.html.erb:7]: indentation.

File: depot/app/views/errors/_error.html.erb [hunk 13]
```diff
@@ -1,10 +1,9 @@
- <% if @order.errors.any? %>
+<% if object.errors.any? %>
   <div id="error_explanation">
-    <h2><%=raw t('errors.error.header', count: @order.errors.count,
-        model: t('activerecord.models.order')) %>.</h2>
-    <p><%= t('errors.error.body') %></p>
+    <h2><%=raw t('errors.error.header', count: object.errors.count,
```
Comment (harsh-ag) [inline depot/app/views/errors/_error.html.erb:3]: indentation

File: depot/app/views/orders/_form.html.erb [hunk 14]
```diff
@@ -1,5 +1,5 @@
 <%= form_for(order) do |f| %>
-  <%= render :partial => 'errors/error' %>
+  <%= render :partial => 'errors/error', locals: { object: order } %>
```
Comment (harsh-ag) [inline depot/app/views/orders/_form.html.erb:2]: use new syntax

File: depot/app/views/users/_form.html.erb [hunk 15]
```diff
@@ -2,7 +2,7 @@
   <fieldset>
     <legend>Enter User Details</legend>
     <%= form_for(user) do |f| %>
-      <%= render :partial => 'errors/users' %>
+      <%= render :partial => 'errors/error', locals: { object: user } %>
```
Comment (harsh-ag) [inline depot/app/views/users/_form.html.erb:5]: new syntax

File: depot/config/environments/development.rb [hunk 16]
```diff
@@ -59,8 +59,8 @@
   config.action_mailer.smtp_settings = {
        :address => "smtp.gmail.com",
        :port => 587,
-       :user_name => "abhikanojiatest@gmail.com",
-       :password => "test@2013",
```
Comment (harsh-ag) [inline depot/config/environments/development.rb:63]: seriously..!!!

File: depot/config/secrets.yml [hunk 17]
```diff
@@ -1,22 +0,0 @@
-# Be sure to restart your server when you modify this file.
```
Comment (harsh-ag) [inline depot/config/secrets.yml:1]: create secrets.yml.tt

File: depot/app/controllers/users_controller.rb [hunk 18]
```diff
@@ -56,10 +55,15 @@ def update
...
-      format.html { redirect_to users_url, notice: 'User was successfully destroyed.' }
-      format.json { head :no_content }
+    if @user.destroy
+      respond_to do |format|
+        format.html { redirect_to users_url, notice: 'User was successfully destroyed.' }
+        format.json { head :no_content }
+      end
+    else
+      respond_to do |format|
+        format.html { redirect_to users_url, notice: @user.errors.full_messages }
```
Comment (harsh-ag) [inline depot/app/controllers/users_controller.rb:65]: use error key

File: depot/.gitignore [hunk 19]
```diff
@@ -21,3 +21,4 @@
 .byebug_history
 /assets/images/*
 /.fuse*
+/config/secrets.yml
```
Comment (harsh-ag) [inline depot/.gitignore:24]: ignore all the yml files and create sample yml files and push them on git

**Review by harsh-ag — CHANGES_REQUESTED** [review]

**Pull request closed without merging**